### PR TITLE
Merge qwen to llama cpp

### PR DESCRIPTION
**Pull request by simonJJJ**
Hi, @ggerganov 

As more and more people begin to use Qwen's open-source models, the influence of Qwen models is growing, especially in China. Many community members are interested in adding support for Qwen models to llama.cpp. To do this, we need to make some changes, which we hope can be merged into the main branch of llama.cpp. In the future, we would be happy to help maintain support for Qwen models in llama.cpp. We sincerely hope that our pull request can be accepted. Thank you.

This PR contains the following features:

- scripts convert Qwen models to gguf format
- add meta info of Qwen models
- inference logic of Qwen models in llama.cpp

```bash
# convert Qwen HF models to gguf fp16 format
python convert-hf-to-gguf.py --outfile qwen7b-chat-f16.gguf --outtype f16 Qwen-7B-Chat

# quantize the model to 4-bits (using q4_0 method)
./build/bin/quantize qwen7b-chat-f16.gguf qwen7b-chat-q4_0.gguf q4_0

# chat with Qwen models
./build/bin/main -m qwen7b-chat-q4_0.gguf -n 512 --color -i -cml -f prompts/chat-with-qwen.txt
```

---

*edits by gg*

- [ ] Qwen 1.8B - cannot convert
- [x] Qwen 7B F16
- [x] Qwen 7B Q8_0
- [x] Qwen 7B Q4_0
- [x] Qwen 14B Q8_0
- [x] Qwen 72B F16
- [x] Qwen 72B Q8_0 - needs fix https://github.com/ggerganov/llama.cpp/pull/4284
- [x] Qwen 72B Q4_0 - needs fix https://github.com/ggerganov/llama.cpp/pull/4284

### CODE REVIEW

**Comment by ggerganov** [comment]
Thank you! I've just finished downloading the models and will now test this PR

**Comment by ggerganov** [comment]
Any ideas how to fix this?

```bash
$ ▶ python3 convert-hf-to-gguf.py models/qwen-1.8b --outfile models/qwen-1.8b/ggml-model-f16.gguf --outtype f16
Loading model: qwen-1.8b
gguf: This GGUF file is for Little Endian only
Set model parameters
Set model tokenizer
Traceback (most recent call last):
  File "/Users/ggerganov/development/github/llama.cpp/convert-hf-to-gguf.py", line 1020, in <module>
    model_instance.set_vocab()
  File "/Users/ggerganov/development/github/llama.cpp/convert-hf-to-gguf.py", line 871, in set_vocab
    tokenizer = AutoTokenizer.from_pretrained(dir_model, trust_remote_code=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers/models/auto/tokenization_auto.py", line 751, in from_pretrained
    tokenizer_class = get_class_from_dynamic_module(class_ref, pretrained_model_name_or_path, **kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers/dynamic_module_utils.py", line 499, in get_class_from_dynamic_module
    return get_class_in_module(class_name, final_module.replace(".py", ""))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers/dynamic_module_utils.py", line 199, in get_class_in_module
    module = importlib.import_module(module_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'transformers_modules.qwen-1'
```

I downloaded the Qwen 1.8B model from here: https://huggingface.co/Qwen/Qwen-1_8B

**Comment by pipixia244** [comment]
> ModuleNotFoundError: No module named 'transformers_modules.qwen-1'
The problem is perhaps the model directory name, which should not contain "_"

**Comment by riverzhou** [comment]
image
Try to load 14B Q8_0 to GPU, failed.

File: llama.cpp [hunk 1]
```diff
@@ -3311,6 +3338,71 @@ static void llm_load_tensors(
...
+                    for (uint32_t i = 0; i < n_layer; ++i) {
+                        const ggml_backend_type backend = int(i) < i_gpu_start ? GGML_BACKEND_CPU : llama_backend_offload; // NOLINT
+                        const ggml_backend_type backend_split = int(i) < i_gpu_start ? GGML_BACKEND_CPU : llama_backend_offload_split; // NOLINT
+
+                        auto & layer = model.layers[i];
+
+                        layer.attn_norm = ml.create_tensor(ctx, tn(LLM_TENSOR_ATTN_NORM, "weight", i), {n_embd}, backend);
+
+                        layer.wqkv = ml.create_tensor(ctx, tn(LLM_TENSOR_ATTN_QKV, "weight", i), {n_embd, n_embd * 3}, backend_split);
+                        layer.bqkv = ml.create_tensor(ctx, tn(LLM_TENSOR_ATTN_QKV, "bias", i),   {n_embd * 3},         backend_split);
```
Comment (ggerganov) [inline llama.cpp:3389]:
```suggestion
                        layer.bqkv = ml.create_tensor(ctx, tn(LLM_TENSOR_ATTN_QKV, "bias", i),   {n_embd * 3},         backend);
```

Should not split bias tensors

**Comment by ggerganov** [comment]
Qwen 72B F16 seems to be working fine:

```bash
make -j && ./main -m ./models/qwen-72b-fast/ggml-model-f16.gguf -p "I believe the meaning of life is" -ngl 1 -s 1 -n 64 --verbose-prompt
```
<details>
<summary>output on M2 Ultra</summary>

```
llm_load_vocab: special tokens definition check successful ( 421/152064 ).
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = qwen
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 152064
llm_load_print_meta: n_merges         = 151387
llm_load_print_meta: n_ctx_train      = 32768
llm_load_print_meta: n_embd           = 8192
llm_load_print_meta: n_head           = 64
llm_load_print_meta: n_head_kv        = 64
llm_load_print_meta: n_layer          = 80
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_gqa            = 1
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-06
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: n_ff             = 49152
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 1000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_yarn_orig_ctx  = 32768
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: model type       = ?B
llm_load_print_meta: model ftype      = mostly F16 (guessed)
llm_load_print_meta: model params     = 72.29 B
llm_load_print_meta: model size       = 134.65 GiB (16.00 BPW) 
llm_load_print_meta: general.name   = Qwen
llm_load_print_meta: BOS token = 151643 '[PAD151643]'
llm_load_print_meta: EOS token = 151643 '[PAD151643]'
llm_load_print_meta: UNK token = 151643 '[PAD151643]'
llm_load_print_meta: LF token  = 148848 'ÄĬ'
llm_load_tensors: ggml ctx size =    0.24 MiB
llm_load_tensors: mem required  = 137884.77 MiB
...................................................................................................
llama_new_context_with_model: n_ctx      = 512
llama_new_context_with_model: freq_base  = 1000000.0
llama_new_context_with_model: freq_scale = 1
llama_new_context_with_model: kv self size  = 1280.00 MiB
llama_build_graph: non-view tensors processed: 2004/2004
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M2 Ultra
ggml_metal_init: picking default device: Apple M2 Ultra
ggml_metal_init: default.metallib not found, loading from source
ggml_metal_init: loading '/Users/ggerganov/development/github/llama.cpp/ggml-metal.metal'
ggml_metal_init: GPU name:   Apple M2 Ultra
ggml_metal_init: GPU family: MTLGPUFamilyApple8 (1008)
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 147456.00 MiB
ggml_metal_init: maxTransferRate               = built-in GPU
llama_new_context_with_model: compute buffer total size = 316.06 MiB
llama_new_context_with_model: max tensor size =  2376.00 MiB
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 110592.00 MiB, offs =            0
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 29674.25 MiB, offs = 113472684032, (140266.88 / 147456.00)
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =  1280.02 MiB, (141546.89 / 147456.00)
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =   313.02 MiB, (141859.91 / 147456.00)

system_info: n_threads = 16 / 24 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 | 

main: prompt: 'I believe the meaning of life is'
main: number of tokens in prompt = 7
    40 -> 'I'
  4411 -> ' believe'
   279 -> ' the'
  7290 -> ' meaning'
   315 -> ' of'
  2272 -> ' life'
   374 -> ' is'

sampling: 
	repeat_last_n = 64, repeat_penalty = 1.100, frequency_penalty = 0.000, presence_penalty = 0.000
	top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
generate: n_ctx = 512, n_batch = 512, n_predict = 64, n_keep = 0


I believe the meaning of life is to live it for others. I’ve tried and failed at many things in my life, but what’s made me happy are those moments when I have helped someone else.
My first wife was a wonderful woman who loved children. She had worked as a nanny（女佣）and always said she wanted a child of her
llama_print_timings:        load time =   18242.98 ms
llama_print_timings:      sample time =      87.74 ms /    64 runs   (    1.37 ms per token,   729.40 tokens per second)
llama_print_timings: prompt eval time =     315.70 ms /     7 tokens (   45.10 ms per token,    22.17 tokens per second)
llama_print_timings:        eval time =   13321.29 ms /    63 runs   (  211.45 ms per token,     4.73 tokens per second)
llama_print_timings:       total time =   13777.97 ms
ggml_metal_free: deallocating
Log end
```
</details>

File: llama.cpp (hunk 1 above)
Comment (riverzhou) [reply llama.cpp:3389]: Tested 14B Q8_0 , is OK.

File: gguf-py/gguf/tensor_mapping.py [hunk 2]
```diff
@@ -161,18 +161,20 @@ class TensorNameMap:
...
             "transformer.h.{bid}.mlp.fc_in",                          # gpt-j
             "language_model.encoder.layers.{bid}.mlp.dense_h_to_4h",  # persimmon
+            "transformer.h.{bid}.mlp.w1",                             # qwen
         ),
 
         # Feed-forward gate
         MODEL_TENSOR.FFN_GATE: (
             "model.layers.{bid}.mlp.gate_proj",  # llama-hf refact
             "layers.{bid}.feed_forward.w1",      # llama-pth
+            "transformer.h.{bid}.mlp.w2",        # qwen
```
Comment (ggerganov) [inline gguf-py/gguf/tensor_mapping.py:171]: Is it possible that `FFN_GATE` and `FFN_UP` are swapped during the conversion?

Comment (ggerganov) [reply gguf-py/gguf/tensor_mapping.py:171]: Hm, based on https://huggingface.co/Qwen/Qwen-7B/blob/e7a368b0774370edec29674e7c51f52fc7663f59/modeling_qwen.py#L572-L577, these are correct

There is something else going on - not sure why the quantum models break

**Comment by CausalLM** [comment]
Tried LLaMAfied Qwen 72B, failed on all quantized, but works well with F16. 
https://github.com/ggerganov/llama.cpp/pull/4283
Using LLaMAfied model: https://huggingface.co/CausalLM/72B-preview
I will try InternLM (LLaMA arch with bias on QKV) later.

**Comment by ggerganov** [comment]
> Tried LLaMAfied Qwen 72B, failed on all quantized, but works well with F16.

Yes, it's also exactly the same with Qwen 72B Chat - F16 works and none of the quantum ones work.

I checked that LLaMA v2 70B works with quantization - it is of similar size.
Maybe it is something related to the bigger vocabulary of Qwen, but I don't understand why it fails only with quantum data

**Comment by ggerganov** [comment]
If the `token_embd` and `output` tensors are not quantized (i.e. leave them in F16) it works.

I am looking more into this and it seems that `token_embd` has both `nan` and `inf` for some of the weights.
This is strange.

**Comment by arivero** [comment]
position embedding failing, perhaps?

**Comment by ggerganov** [comment]
Wow, we have a bug in the quantization code - integer overflow when using multiple threads:

https://github.com/ggerganov/llama.cpp/blob/8d6d9f033b8101f929e445cf45b39e1557ca7934/llama.cpp#L7657-L7671

Some of these `auto`'s end up 32-bit integers instead of 64-bit.
Took me quite some time to find this!

Will push a fix shortly

**Comment by ggerganov** [comment]
We just need to fix Qwen-1.8B conversion now and we can merge.

Any ideas what this error means?
I'm very interested in trying this model

**Comment by simonJJJ** [comment]
@ggerganov thank you so much. I have a very bad fever since Friday night.

**Comment by zhangp365** [comment]
> > From the log, I can find some chat template information, but not all. I also searched on Hugging Face and GitHub, but I couldn't find the complete chat template. Could you please provide the chat template?
> > ```
> > Reverse prompt: '<|im_start|>user
> > '
> > sampling:
> >         repeat_last_n = 64, repeat_penalty = 1.100, frequency_penalty = 0.000, presence_penalty = 0.000
> >         top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
> >         mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
> > generate: n_ctx = 512, n_batch = 512, n_predict = 512, n_keep = 19
> > 
> > 
> > == Running in interactive mode. ==
> >  - Press Ctrl+C to interject at any time.
> >  - Press Return to return control to LLaMa.
> >  - To return control without starting a new line, end your input with '/'.
> >  - If you want to submit another line, end your input with '\'.
> > 
> > <|im_start|>system
> > You are a helpful assistant.<|im_end|>
> > ```
> 
> See make_context [link](https://huggingface.co/Qwen/Qwen-7B/blob/main/modeling_qwen.py#L1126).

I got it, thank you very much.

**Comment by svcvit** [comment]
Unable to support Qwen-VL-Chat.  `python convert-hf-to-gguf.py --outfile qwenvl-chat-f16.gguf --outtype f16 Qwen-VL-Chat` 

```
………………
blk.31.attn_qkv.bias, n_dims = 1, torch.bfloat16 --> float32
blk.31.attn_output.weight, n_dims = 2, torch.bfloat16 --> float32
blk.31.ffn_norm.weight, n_dims = 1, torch.bfloat16 --> float32
blk.31.ffn_up.weight, n_dims = 2, torch.bfloat16 --> float32
blk.31.ffn_gate.weight, n_dims = 2, torch.bfloat16 --> float32
blk.31.ffn_down.weight, n_dims = 2, torch.bfloat16 --> float32
output_norm.weight, n_dims = 1, torch.bfloat16 --> float32
Can not map tensor 'transformer.visual.positional_embedding'
```

**Comment by dgo2dance** [comment]
> ```shell
> $ ▶ python3 convert-hf-to-gguf.py models/qwen-1.8b --outfile models/qwen-1.8b/ggml-model-f16.gguf --outtype f
> ```

So, can the qwen1.8B be successfully converted now

**Comment by hobbytp** [comment]
@simonJJJ please also update the "Supported models" list in the llama.cpp README on the first page.

**Comment by simonJJJ** [comment]
> @simonJJJ please also update the "Supported models" list in the llama.cpp README on the first page.

will do it

**Comment by zhangjiekui** [comment]
Yes, it can run. But the generating result seems not so good. I tried 14B gguf, but get so many reptitions , even likes '/n/n/n‘.........for  tens time.   So there is sth must be corrected or improved. Thanks

**Comment by nlp4whp** [comment]
> Yes, it can run. But the generating result seems not so good. I tried 14B gguf, but get so many reptitions , even likes '/n/n/n‘.........for tens time. So there is sth must be corrected or improved. Thanks

If I add suffix like `\nAI:` in query (`User:{query}\nAI:`), it will return normal result most time; I tried `qwen-7B-chat + q4_0`

**Comment by riverzhou** [comment]
> Yes, it can run. But the generating result seems not so good. I tried 14B gguf, but get so many reptitions , even likes '/n/n/n‘.........for tens time. So there is sth must be corrected or improved. Thanks

Same problem.

**Comment by AnLoffredo** [comment]
I actually have a similar problem. I have \n set to return control, but the model will generate !!! constantly, even when not appropriate.  It's weird because all three have the same logit value, and the correct punctuation usually follows.

`Hey!!!, how are you!!!?`

It's always in a set of three. I assumed it was something I was doing wrong but now I'm not entirely sure.

**Comment by 2213601279** [comment]
python3 convert-hf-to-gguf.py --outfile qwen14b-chat-f16.gguf --outtype f16 ../../LLM/Qwen-14B-Chat/
Loading model: Qwen-14B-Chat
gguf: This GGUF file is for Little Endian only
Set model parameters
Set model tokenizer
gguf: Adding 151387 merge(s).
gguf: Setting special token type bos to 151643
gguf: Setting special token type eos to 151643
gguf: Setting special token type unk to 151643
Exporting model to 'qwen14b-chat-f16.gguf'
gguf: loading model part 'model-00001-of-00015.safetensors'
Traceback (most recent call last):
  File "/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py", line 1027, in <module>
    model_instance.write()
  File "/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py", line 126, in write
    self.write_tensors()
  File "/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py", line 925, in write_tensors
    model_kv = dict(self.get_tensors())
  File "/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py", line 60, in get_tensors
    ctx = cast(ContextManager[Any], safe_open(self.dir_model / part_name, framework="pt", device="cpu"))
RuntimeError: unable to mmap 2047396712 bytes from file <../../LLM/Qwen-14B-Chat/model-00001-of-00015.safetensors>: Cannot allocate memory (12)
@simonJJJ what can i do with this ? in my PC  memory is 64GB ,i think it's enough

**Comment by 2213601279** [comment]
> python3 convert-hf-to-gguf.py --outfile qwen14b-chat-f16.gguf --outtype f16 ../../LLM/Qwen-14B-聊天/ 加载型号：Qwen-14B-Chat gguf：此 GGUF 文件仅适用于 Little Endian 设置模型参数 设置模型分词器 gguf：添加151387合并。 gguf：设置特殊令牌类型 bos 为 151643 gguf：将特殊令牌类型 eos 设置为151643 gguf：将特殊令牌类型 unk 设置为 151643 导出模型到“qwen14b-chat-f16.gguf” gguf：加载模型部件“model-00001-of-00015.safetensors” 回溯（最近一次调用最后一次）： 文件 “/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py”，第 1027 行，在 model_instance.write（） 中文件“/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py”，第 126 行，写入 self.write_tensors（） 文件“/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py”，第 925 行，write_tensors model_kv = dict（self.get_tensors（）） 文件“/mnt/e/LLMInf/llama.cpp/convert-hf-to-gguf.py”，第 60 行，get_tensors ctx = cast（ContextManager[Any]， safe_open（self.dir_model / part_name， framework=“pt”， device=“cpu”））RuntimeError：无法从文件< mmap 2047396712 字节。/../LLM/Qwen-14B-Chat/model-00001-of-00015.safetensors>：无法分配内存 （12） 我能用这个做什么？在我的 PC 内存中是 64GB，我认为这已经足够了

Qwen-14B-Chat this model come from modelscope platfrom

**Comment by yanshuaibupt** [comment]
I generated qwen14b-chat-f16.gguf by the script, but when I set the qwen in localai as the backend, I got the following problem:
image

**Comment by renraeldab** [comment]
It can run without errors but I think something is wrong with the chat completion. It always outputs a lot of "\n" no matter how I change the history messages. `transformers.AutoModelForCausalLM.chat()` for Qwen parses messages in its own way, but I am not sure if `Llama.create_chat_completion()` is doing this right.

```
llama_print_timings:        load time =     396.49 ms
llama_print_timings:      sample time =     548.76 ms /   717 runs   (    0.77 ms per token,  1306.59 tokens per second)
llama_print_timings: prompt eval time =     396.43 ms /    15 tokens (   26.43 ms per token,    37.84 tokens per second)
llama_print_timings:        eval time =   39471.61 ms /   716 runs   (   55.13 ms per token,    18.14 tokens per second)
llama_print_timings:       total time =   49647.80 ms /   731 tokens
INFO:     ::1:54012 - "POST /v1/chat/completions HTTP/1.1" 200 OK
```

```
{'id': 'chatcmpl-9cc8ef09-405a-4ac9-b1bc-4bf3eca02f16', 'object': 'chat.completion', 'created': 1711962634, 'model': 'Qwen-1_8B_Chat/ggml-model-Q4_K_M.gguf', 'choices': [{'index': 0, 'message': {'content': '    <br><br>\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n', 'role': 'assistant'}, 'finish_reason': 'stop'}], 'usage': {'prompt_tokens': 15, 'completion_tokens': 716, 'total_tokens': 731}}
```

For comparison, this is the result of completion.

```
llama_print_timings:        load time =     396.49 ms
llama_print_timings:      sample time =      13.93 ms /    16 runs   (    0.87 ms per token,  1148.60 tokens per second)
llama_print_timings: prompt eval time =     160.45 ms /     4 tokens (   40.11 ms per token,    24.93 tokens per second)
llama_print_timings:        eval time =     761.05 ms /    15 runs   (   50.74 ms per token,    19.71 tokens per second)
llama_print_timings:       total time =    1149.10 ms /    19 tokens
INFO:     ::1:54086 - "POST /v1/completions HTTP/1.1" 200 OK
```

```
{'id': 'cmpl-c9a4c974-d8dc-48b3-82cf-8fd73e9fb1a2', 'object': 'text_completion', 'created': 1711963517, 'model': 'Qwen-1_8B_Chat/ggml-model-Q4_K_M.gguf', 'choices': [{'text': '\n\n\nHello! How may I assist you today?', 'index': 0, 'logprobs': None, 'finish_reason': 'length'}], 'usage': {'prompt_tokens': 5, 'completion_tokens': 16, 'total_tokens': 21}}
```

**Pull request merged**